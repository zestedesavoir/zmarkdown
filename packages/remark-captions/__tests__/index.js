import dedent from 'dedent'
import unified from 'unified'
import reParse from 'remark-parse'
import stringify from 'rehype-stringify'
import remark2rehype from 'remark-rehype'
import remarkStringify from 'remark-stringify'

import remarkCaptions from '../src/'
import remarkGridTables from '../../remark-grid-tables/src'


const render = (text, config) => unified()
  .use(reParse, {
    gfm: true,
    commonmark: false,
    footnotes: true,
    /* sets list of known blocks to nothing, otherwise <h3>hey</h3> would become
    &#x3C;h3>hey&#x3C;/h3> instead of <p>&#x3C;h3>hey&#x3C;/h3></p> */
    blocks: [],
  })
  .use(remarkGridTables)
  .use(remarkCaptions, config)
  .use(remark2rehype)
  .use(stringify)
  .processSync(text)


const renderToMarkdown = (text, config) => unified()
  .use(reParse)
  .use(remarkStringify)
  .use(remarkCaptions, config)
  .processSync(text)


test('code', () => {
  const {contents} = render(dedent`
    ## Code

    Normal code

    \`\`\`python
    print('bla')
    \`\`\`

    With Legend

    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1

    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1
    break


    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1 *em* **strong \`code\`** end
    bla

    \`\`\`python
    print('bla2')
    \`\`\`
    Code: figcapt1
    Code: bis
  `)
  expect(contents).toMatchSnapshot()
})

test('#101', () => {
  const {contents} = render(dedent`
  ## code followed by non text par

  \`\`\`
  foo
  \`\`\`

  [](#)`, {external: {code: 'Code:'}})
  expect(contents).toMatchSnapshot()
})

test('custom-table', () => {
  const {contents} = render(dedent`
    ## Table

    Normal table


    head1| head2
    -----|------
    bla|bla


    With Legend


    head1| head2
    -----|------
    bla|bla
    CustomTable: figcapt1


    head1| head2
    -----|------
    bla|bla
    CustomTable: figcapt1
    CustomTable: bis
    `,
  {external: {table: 'CustomTable:'}}
  )
  expect(contents).toMatchSnapshot()
})

test('gridtables', () => {
  const {contents} = render(dedent`
    +----+----+----+
    | a  |  b | c  |
    +====+====+====+
    | 1  |2   |3   |
    +----+----+----+
    Table: bla bla`,
  {external: {gridTable: 'Table:'}}
  )
  expect(contents).toMatchSnapshot()
})

test('caption without block', () => {
  const {contents} = render(dedent`
      \`\`\`python
      print('bla')
      \`\`\`

      \`\`\`python hl_lines=1,2
      print('bla')
      print('bla')
      print('bla')
      \`\`\`

      \`\`\`
      a code without lang
      \`\`\`
    `, {external: {gridTable: 'Table:'}}
  )
  expect(contents).toMatchSnapshot()
})

test('quotation', () => {
  const {contents} = render(dedent`
    (for convenience, · are replaced with
    simple single spaces in the tests)

    ## Blockquote

    Empty blockquote

    >


    Normal blockquote

    > My citation

    With Legend

    > My citation
    Source: figcapt1

    With Legend without Source:

    > My citation
    : figcapt2

    haha

    > > > Foo··
    > > > Foo
    > > Source: fc1
    > >
    > > Bar··
    > > Bar··
    > Source: fc2
    >
    > Baz
    > Baz
    Source: fc3
  `.replace(/·/g, ' '))
  expect(contents).toMatchSnapshot()
})

test('captions should break lists', () => {
  const {contents} = render(dedent`
    > * Bar
    Source: fc2

    > * Baz
    >     * Baz
    Source: fc3
  `.replace(/·/g, ' '))
  expect(contents).toMatchSnapshot()
})

test('table', () => {
  const {contents} = render(dedent`
    ## Table

    Normal table


    head1| head2
    -----|------
    bla|bla


    With Legend


    head1| head2
    -----|------
    bla|bla
    Table: figcapt1
  `)
  expect(contents).toMatchSnapshot()
})

test('external legend: two legends', () => {
  const {contents} = render(dedent`
    head1| head2
    -----|------
    bla|bla
    Table: figcapt1
    Table: bis
  `)
  expect(contents).toMatchSnapshot()
})

test('internal legend: two legends', () => {
  const {contents} = render(dedent`
    Should only keep the 1st

    > My citation
    Source: first capt
    Source: last capt··
    2nd line

    noop

    > foo
    > bar
    > baz
    > qux
    Source: **first**
    b*a*r
    Source: This is **the real** \`source\`

    noop
  `.replace(/·/g, ' '))
  expect(contents).toMatchSnapshot()
})

test('legend in paragraph', () => {
  const {contents} = render(dedent`
    foo

    ![]()
    Figure: 1 this is parsed as legend

    baz

    ![]()
    aFigure: 2 this is displayed as text

    ![alt 2b](https://zestedesavoir.com/static/images/home-clem.4a2f792744c9.png)
    this is displayed as text

    foo
    ![]()
    Figure: 3 this is a legend

    ![]()
    Figure: 4 this is a legend, remainder of the paragraph goes into
    the
    legend

    Figure: 5 this is a text with and image
    ![]()
    in the middle

    Figure: 6 displayed as text
    ![]()
    Figure: 7 is a legend

    ![]()
    Figure: 8 is a legend.

    Figure: 9 this is a text.
  `.replace(/·/g, ' '))
  expect(contents).toMatchSnapshot()
})

test('compiles to markdown when no caption', () => {
  const md = dedent`
    foo

    ![](img)
    Figure: 1 this is parsed as legend

    baz

    ![](img)
    aFigure: 2 this is displayed as text

    ![alt 2b](https://zestedesavoir.com/static/images/home-clem.4a2f792744c9.png)
    this is displayed as text
  `
  const {contents} = renderToMarkdown(md)
  expect(contents).toMatchSnapshot()

  const contents1 = renderToMarkdown(md).contents
  const contents2 = renderToMarkdown(contents1).contents

  expect(contents1).toBe(contents2)
})
test('Two images in figure', () => {
  const md = dedent`
    ![alt text1](img1.png)
    ![alt text2](img2.png)
    Figure: Common text
  `
  const {contents} = render(md)
  expect(contents).toMatchSnapshot()
})
test('compiles to markdown when at least 1 block caption', () => {
  const md = dedent`
    foo
    ![](img)
    Figure: 3 this is a legend

    ![](img)
    Figure: 4 this is a legend, remainder of the paragraph goes into
    the
    legend

    Figure: 5 this is a text with and image
    ![](<title> img)
    in the middle

    Figure: 6 displayed as text
    ![](img)
    Figure: 7 is a legend

    ![](img)
    Figure: 8 is a legend.

    Figure: 9 this is a text.

    > My citation
    Source: first capt
    Source: last capt··
    2nd line

    noop

    > foo
    > bar
    > baz
    > qux
    Source: **first**
    b*a*r
    Source: This is **the real** \`source\`

    noop

    > foo
    > bar
    > baz
    > qux
    Source: **first**
    b*a*r
    Source: This is **the real** \`source\`

  `
  const {contents} = renderToMarkdown(md)
  expect(contents).toMatchSnapshot()

  const contents1 = renderToMarkdown(md).contents
  const contents2 = renderToMarkdown(contents1).contents

  expect(contents1).toBe(contents2)
})

test('compiles to markdown when at least 1 innerLegend caption', () => {
  const md = dedent`
    ## Code

    Normal code

    \`\`\`python
    print('bla')
    \`\`\`

    With Legend

    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1

    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1
    break

    \`\`\`python
    print('bla')
    \`\`\`
    Code: figcapt1 *em* **strong \`code\`** end
    bla

    \`\`\`python
    print('bla2')
    \`\`\`
    Code: figcapt1
    Code: bis
  `
  const {contents} = renderToMarkdown(md)
  expect(contents).toMatchSnapshot()

  const contents1 = renderToMarkdown(md).contents
  const contents2 = renderToMarkdown(contents1).contents

  expect(contents1).toBe(contents2)
})
