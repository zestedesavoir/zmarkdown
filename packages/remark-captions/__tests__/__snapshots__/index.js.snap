// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#101 1`] = `
"<h2>code followed by non text par</h2>
<pre><code>foo
</code></pre>
<p><a href=\\"#\\"></a></p>"
`;

exports[`Two images in figure 1`] = `"<figure><img src=\\"img1.png\\" alt=\\"alt text1\\"><img src=\\"img2.png\\" alt=\\"alt text2\\"><figcaption>Common text</figcaption></figure>"`;

exports[`caption without block 1`] = `
"<pre><code class=\\"language-python\\">print('bla')
</code></pre>
<pre><code class=\\"language-python\\">print('bla')
print('bla')
print('bla')
</code></pre>
<pre><code>a code without lang
</code></pre>"
`;

exports[`captions should break lists 1`] = `
"<figure><blockquote>
<ul>
<li>Bar</li>
</ul>
</blockquote><figcaption>fc2</figcaption></figure>
<figure><blockquote>
<ul>
<li>Baz
<pre><code>* Baz
</code></pre>
</li>
</ul>
</blockquote><figcaption>fc3</figcaption></figure>"
`;

exports[`code 1`] = `
"<h2>Code</h2>
<p>Normal code</p>
<pre><code class=\\"language-python\\">print('bla')
</code></pre>
<p>With Legend</p>
<figure><pre><code class=\\"language-python\\">print('bla')
</code></pre><figcaption>figcapt1</figcaption></figure>
<figure><pre><code class=\\"language-python\\">print('bla')
</code></pre><figcaption>figcapt1</figcaption></figure>
<p>break</p>
<figure><pre><code class=\\"language-python\\">print('bla')
</code></pre><figcaption>figcapt1 <em>em</em> <strong>strong <code>code</code></strong> end</figcaption></figure>
<p>bla</p>
<figure><pre><code class=\\"language-python\\">print('bla2')
</code></pre><figcaption>figcapt1</figcaption></figure>
<p>Code: bis</p>"
`;

exports[`compiles to markdown when at least 1 block caption 1`] = `
"foo

![](img)
Figure: 3 this is a legend

![](img)
Figure: 4 this is a legend, remainder of the paragraph goes into
the
legend

Figure: 5 this is a text with and image
!\\\\[](<title> img)
in the middle

Figure: 6 displayed as text

![](img)
Figure: 7 is a legend

![](img)
Figure: 8 is a legend.

Figure: 9 this is a text.

> My citation
> Source: first capt
Source: last capt··
2nd line

noop

> foo
> bar
> baz
> qux
> Source: **first**
> b_a_r
Source: This is **the real** \`source\`

noop

> foo
> bar
> baz
> qux
> Source: **first**
> b_a_r
Source: This is **the real** \`source\`
"
`;

exports[`compiles to markdown when at least 1 innerLegend caption 1`] = `
"## Code

Normal code

\`\`\`python
print('bla')
\`\`\`

With Legend

\`\`\`python
print('bla')
\`\`\`
Code: figcapt1

\`\`\`python
print('bla')
\`\`\`
Code: figcapt1

break

\`\`\`python
print('bla')
\`\`\`
Code: figcapt1 _em_ **strong \`code\`** end

bla

\`\`\`python
print('bla2')
\`\`\`
Code: figcapt1

Code: bis
"
`;

exports[`compiles to markdown when no caption 1`] = `
"foo

![](img)
Figure: 1 this is parsed as legend

baz

![](img)
aFigure: 2 this is displayed as text

![alt 2b](https://zestedesavoir.com/static/images/home-clem.4a2f792744c9.png)
this is displayed as text
"
`;

exports[`custom-table 1`] = `
"<h2>Table</h2>
<p>Normal table</p>
<table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table>
<p>With Legend</p>
<figure><table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table><figcaption>figcapt1</figcaption></figure>
<figure><table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table><figcaption>figcapt1</figcaption></figure>
<p>CustomTable: bis</p>"
`;

exports[`external legend: two legends 1`] = `
"<figure><table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table><figcaption>figcapt1</figcaption></figure>
<p>Table: bis</p>"
`;

exports[`gridtables 1`] = `"<figure><table><thead><tr><th colspan=\\"1\\" rowspan=\\"1\\"><p>a</p></th><th colspan=\\"1\\" rowspan=\\"1\\"><p>b</p></th><th colspan=\\"1\\" rowspan=\\"1\\"><p>c</p></th></tr></thead><tbody><tr><td colspan=\\"1\\" rowspan=\\"1\\"><p>1</p></td><td colspan=\\"1\\" rowspan=\\"1\\"><p>2</p></td><td colspan=\\"1\\" rowspan=\\"1\\"><p>3</p></td></tr></tbody></table><figcaption>bla bla</figcaption></figure>"`;

exports[`internal legend: two legends 1`] = `
"<p>Should only keep the 1st</p>
<figure><blockquote>
<p>My citation
Source: first capt</p>
</blockquote><figcaption>last capt<br>
2nd line</figcaption></figure>
<p>noop</p>
<figure><blockquote>
<p>foo
bar
baz
qux
Source: <strong>first</strong>
b<em>a</em>r</p>
</blockquote><figcaption>This is <strong>the real</strong> <code>source</code></figcaption></figure>
<p>noop</p>"
`;

exports[`legend in paragraph 1`] = `
"<p>foo</p>
<figure><img src=\\"\\"><figcaption>1 this is parsed as legend</figcaption></figure>
<p>baz</p>
<p><img src=\\"\\">
aFigure: 2 this is displayed as text</p>
<p><img src=\\"https://zestedesavoir.com/static/images/home-clem.4a2f792744c9.png\\" alt=\\"alt 2b\\">
this is displayed as text</p>
<p>foo</p>
<figure><img src=\\"\\"><figcaption>3 this is a legend</figcaption></figure>
<figure><img src=\\"\\"><figcaption>4 this is a legend, remainder of the paragraph goes into
the
legend</figcaption></figure>
<p>Figure: 5 this is a text with and image
<img src=\\"\\">
in the middle</p>
<p>Figure: 6 displayed as text</p>
<figure><img src=\\"\\"><figcaption>7 is a legend</figcaption></figure>
<figure><img src=\\"\\"><figcaption>8 is a legend.</figcaption></figure>
<p>Figure: 9 this is a text.</p>"
`;

exports[`quotation 1`] = `
"<p>(for convenience,   are replaced with
simple single spaces in the tests)</p>
<h2>Blockquote</h2>
<p>Empty blockquote</p>
<blockquote>
</blockquote>
<p>Normal blockquote</p>
<blockquote>
<p>My citation</p>
</blockquote>
<p>With Legend</p>
<figure><blockquote>
<p>My citation</p>
</blockquote><figcaption>figcapt1</figcaption></figure>
<p>With Legend without Source:</p>
<blockquote>
<p>My citation
: figcapt2</p>
</blockquote>
<p>haha</p>
<figure><blockquote>
<figure><blockquote>
<figure><blockquote>
<p>Foo<br>
Foo</p>
</blockquote><figcaption>fc1</figcaption></figure>
<p>Bar<br>
Bar<br>
</p>
</blockquote><figcaption>fc2</figcaption></figure>
<p>Baz
Baz</p>
</blockquote><figcaption>fc3</figcaption></figure>"
`;

exports[`table 1`] = `
"<h2>Table</h2>
<p>Normal table</p>
<table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table>
<p>With Legend</p>
<figure><table>
<thead>
<tr>
<th>head1</th>
<th>head2</th>
</tr>
</thead>
<tbody>
<tr>
<td>bla</td>
<td>bla</td>
</tr>
</tbody>
</table><figcaption>figcapt1</figcaption></figure>"
`;
